<html><body><div><div class="section">
              <h1>Katana 0.1.1</h1>

              


<p>Command-line tool to soft-clip reads based on primer locations.</p><p>








======</p><p>Katana</p><p>======</p><p>Command-line tool to soft-clip reads from amplicon-based sequence based on</p><p>specified primer locations.</p><p>.. image:: https://travis-ci.org/umich-brcf-bioinf/Katana.svg?branch=develop</p><p>    :target: https://travis-ci.org/umich-brcf-bioinf/Katana</p><p>    :alt: Build Status</p><p>.. image:: https://coveralls.io/repos/github/umich-brcf-bioinf/Katana/badge.svg?branch=develop</p><p>    :target: https://coveralls.io/github/umich-brcf-bioinf/Katana?branch=develop</p><p>    :alt: Coverage Status</p><p>The official repository is at:</p><p>https://github.com/umich-brcf-bioinf/Katana</p><p>--------</p><p>Overview</p><p>--------</p><p>In amplicon-based target panel sequencing, regions-of-interest are amplified by</p><p>specific pairs of primers; consequently the regions-of-interest typically</p><p>always start and end with these primer sequences, sequences which match the</p><p>reference sequence exactly and do not reflect the actual sample sequence. In</p><p>some panel designs, the amplicons may be tiled such that an amplicon of one</p><p>region of interest may overlap the primer region of different amplicon. In this</p><p>arrangement, the overlapping regions should enable detection of variants that</p><p>fall within that primer region. However, the presence of the primer sequences</p><p>will typically overwhelm the signature of true, low-frequency variants.</p><p>Katana matches each read to its corresponding primer pair based on start</p><p>position of the read. Katana then soft-clips the primer region from the edge of</p><p>the read sequence, rescuing the signal of true variants measured by overlapping</p><p>amplicons. The output is conceptually similar to hard-clipping the primers from</p><p>the original FASTQ reads based on sequence identity but with the advantage that</p><p>retaining the primers during alignment improves alignment quality.</p><p>::</p><p>  amplicon A           [ primerREGION-OF-INTERESTprimer ]</p><p>  amplicon B                       [ primerREGION-OF-INTERESTprimer ]</p><p>  input read1 sequence:  TGCATGAGTCTGATCTAGGTAGTTGACGTC</p><p>  input read2 sequence:              ATCTAGGTAGTTGACGTCAGATAATGCAGC</p><p>  output read1 sequence: tgcatgAGTCTGATCTAGGTAGTTgacgtc (clipped amplicon A primers)</p><p>  output read2 sequence:             atctagGTAGTTGACGTCAGATAAtgcagc (clipped amplicon B primers)</p><p>  (lowercase = soft-clipped)</p><p>Tags are added to each output read to help explain how it was modified:</p><p> - X0 : associated primer id</p><p> - X1 : original cigar string</p><p> - X2 : original reference start</p><p> - X3 : original reference_end (informational; useful for reverse reads)</p><p> - X4 : why read would be excluded (appears only if --preserve_all_alignments)</p><p>Katana assumes that:</p><p> - input bam is indexed</p><p> - primers come in sense-antisense pairs</p><p> - primer pairs are on the same chromosome</p><p> - primer chromsomes match the bam regions</p><p> - primer file is tab separated; the header line includes the following fields:</p><p>    * Customer TargetID</p><p>    * Chr</p><p>    * Sense Start</p><p>    * Antisense Start</p><p>    * Sense Sequence</p><p>    * Antisense Sequence</p><p> - primer file sense and antisense start are specified in 1-based coordinates</p><p>-----------</p><p>Quick Start</p><p>-----------</p><p>1. **Install Katana (see INSTALL.rst):**</p><p>::</p><p>  $ pip install katana</p><p>2. **Get the examples directory:**</p><p>::</p><p>  $ git clone https://github.com/umich-brcf-bioinf/Katana</p><p>3. **Run Katana:**</p><p>::</p><p>  $ katana Katana/examples/primers.txt Katana/examples/chr10.pten.bam clipped.bam</p><p>This will read chr10.pten.bam and produce clipped.bam which contains reads</p><p>adjusted to soft-clip (exclude) their respective primer regions. Unmapped reads</p><p>or reads which do not match a known primer are excluded.</p><p>-----------</p><p>Katana help</p><p>-----------</p><p>::</p><p>  $ katana --help  </p><p>  usage: katana primer_manifest input_bam output_bam  </p><p>  Match each alignment in input BAM to primer, softclipping the primer region.  </p><p>  positional arguments:</p><p>    primer_manifest       path to primer manifest (tab-separated text)</p><p>    input_bam             path to input BAM</p><p>    output_bam            path to output BAM    </p><p>  optional arguments:</p><p>    -h, --help            show this help message and exit</p><p>    -V, --version         show program's version number and exit</p><p>    --preserve_all_alignments</p><p>                          Preserve all incoming alignments (even if they are </p><p>                          unmapped, cannot be matched with primers, result in </p><p>                          invalid CIGARs, etc.)</p><p>====</p><p>Email bfx-katana@umich.edu for support and questions.</p><p>UM BRCF Bioinformatics Core</p><p>Changelog</p><p>=========</p><p>0.1.1 (XX/XX/XXXX)</p><p>-----------------</p><p>- Fixed problems in BAM output:</p><p>   - Corrected next reference in paired reads</p><p>   - Excludes reads where CIGAR is entirely clipped</p><p>   - Unpairs reads which had no mate in input</p><p>- Added BAM tags to excluded reads (useful when --preserving_all_reads)</p><p>- Adjusted to improve performance (about 6x faster)</p><p>- Added support for pip install</p><p>- Added functional tests</p><p>- Added support for travis CI</p><p>- Added support for Python3</p><p>- Added support for pysam 0.8.3</p><p>0.1 (1/28/2016)</p><p>--------------</p><p>- Initial Release</p><p>Katana is written and maintained by the University of Michigan </p><p>BRCF Bioinformatic Core; individual contributors include:</p><p>- Chris Gates</p><p>- Peter Ulintz

</p><a name="downloads">Â </a>


<ul class="nodot">
  <li><strong>Downloads (All Versions):</strong></li>
  <li>
    <span>0</span> downloads in the last day
  </li>
  <li>
    <span>75</span> downloads in the last week
  </li>
  <li>
    <span>633</span> downloads in the last month
  </li>
</ul>









            </div>


          </div></body></html>