<html><body><div><article class="markdown-body entry-content" itemprop="text"><h1><a id="user-content-pygeno-a-python-package-for-precision-medicine" class="anchor" href="#pygeno-a-python-package-for-precision-medicine" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" role="img" version="1.1" viewbox="0 0 16 16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"/></svg></a>pyGeno: a Python Package for Precision Medicine</h1>
<a href="https://camo.githubusercontent.com/04120dff48daf616ee76b8add0b43bba02180fea/687474703a2f2f62696f696e666f2e697269632e63612f7e64616f756461742f707947656e6f2f5f7374617469632f6c6f676f2e706e67" target="_blank"><img alt="pyGeno's logo" src="https://camo.githubusercontent.com/04120dff48daf616ee76b8add0b43bba02180fea/687474703a2f2f62696f696e666f2e697269632e63612f7e64616f756461742f707947656e6f2f5f7374617469632f6c6f676f2e706e67" data-canonical-src="http://bioinfo.iric.ca/~daoudat/pyGeno/_static/logo.png"/></a>
<p>Even though more and more research focuses on Personalized/Precision Medicine, treatments that are specially tailored to the patient, pyGeno is (to our knowledge) the only tool available that will gladly build your specific genomes for you and you give an easy access to them.</p>
<p>I really want pyGeno to help as many people as possible do as much life saving research as possible. Therefore, if you find any bug, please fill in a github issue, or even better, fix it and send me a pull request :). I'll gladly mention your contribution on the website. That also goes for the doc.</p>
<p>If you are using pyGeno please mention it to the rest of the universe by including a link to: <a href="https://github.com/tariqdaouda/pyGeno">https://github.com/tariqdaouda/pyGeno</a></p>
<p>pyGeno is developed by <a href="http://wwww.tariqdaouda.com">Tariq Daouda</a> at the <em>Institute for Research in Immunology and Cancer</em> (<a href="http://www.iric.ca">IRIC</a>), its logo is the work of the freelance designer <a href="http://www.sawssankaddoura.com">Sawssan Kaddoura</a>.
For the latest news about pyGeno, you can follow me on twitter <a href="https://www.twitter.com/tariqdaouda">@tariqdaouda</a>.</p>
<p>Click here for The <a href="http://pygeno.iric.ca/">full documentation</a>.</p>
<p>For the latest news about pyGeno, you can follow me on twitter <a href="https://www.twitter.com/tariqdaouda">@tariqdaouda</a>.</p>
<a name="user-content-installation"/>
<h2><a id="user-content-installation" class="anchor" href="#installation" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" role="img" version="1.1" viewbox="0 0 16 16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"/></svg></a>Installation:</h2>
<div class="highlight highlight-source-python"><pre>pip install pyGeno <span class="pl-c">#for the latest stable version</span></pre></div>
<p>For the latest developments:</p>
<div class="highlight highlight-source-shell"><pre>git clone https://github.com/tariqdaouda/pyGeno.git
<span class="pl-c1">cd</span> pyGeno
python setup.py develop</pre></div>
<a name="user-content-a-brief-introduction"/>
<h2><a id="user-content-a-brief-introduction" class="anchor" href="#a-brief-introduction" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" role="img" version="1.1" viewbox="0 0 16 16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"/></svg></a>A brief introduction</h2>
<p>pyGeno is a personal bioinformatic database that runs directly into python, on your laptop and does not depend
upon any REST API. pyGeno is here to make extracting data such as gene sequences a breeze, and is designed to
be able cope with huge queries. The most exciting feature of pyGeno, is that it allows to work with seamlessly with both reference and <strong>Personalized Genomes</strong>.</p>
<p>Personalized Genomes, are custom genomes that you create by combining a reference genome, sets of polymorphisms and an optional filter.
pyGeno will take care of applying the filter and inserting the polymorphisms at their right place, so you get
direct access to the DNA and Protein sequences of your patients.</p>
<div class="highlight highlight-source-python"><pre><span class="pl-k">from</span> pyGeno.Genome <span class="pl-k">import</span> <span class="pl-k">*</span>

g <span class="pl-k">=</span> Genome(<span class="pl-v">name</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">"</span>GRCh37.75<span class="pl-pds">"</span></span>)
prot <span class="pl-k">=</span> g.get(Protein, <span class="pl-v">id</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span>ENSP00000438917<span class="pl-pds">'</span></span>)[<span class="pl-c1">0</span>]
<span class="pl-c">#print the protein sequence</span>
<span class="pl-c1">print</span> prot.sequence
<span class="pl-c">#print the protein's gene biotype</span>
<span class="pl-c1">print</span> prot.gene.biotype
<span class="pl-c">#print protein's transcript sequence</span>
<span class="pl-c1">print</span> prot.transcript.sequence

<span class="pl-c">#fancy queries</span>
<span class="pl-k">for</span> exon <span class="pl-k">in</span> g.get(Exon, {<span class="pl-s"><span class="pl-pds">"</span>CDS_start &gt;<span class="pl-pds">"</span></span>: x1, <span class="pl-s"><span class="pl-pds">"</span>CDS_end &lt;=<span class="pl-pds">"</span></span> : x2, <span class="pl-s"><span class="pl-pds">"</span>chromosome.number<span class="pl-pds">"</span></span> : <span class="pl-s"><span class="pl-pds">"</span>22<span class="pl-pds">"</span></span>}) :
        <span class="pl-c">#print the exon's coding sequence</span>
        <span class="pl-c1">print</span> exon.<span class="pl-c1">CDS</span>
        <span class="pl-c">#print the exon's transcript sequence</span>
        <span class="pl-c1">print</span> exon.transcript.sequence

<span class="pl-c">#You can do the same for your subject specific genomes</span>
<span class="pl-c">#by combining a reference genome with polymorphisms</span>
g <span class="pl-k">=</span> Genome(<span class="pl-v">name</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">"</span>GRCh37.75<span class="pl-pds">"</span></span>, <span class="pl-v">SNPs</span> <span class="pl-k">=</span> [<span class="pl-s"><span class="pl-pds">"</span>STY21_RNA<span class="pl-pds">"</span></span>], <span class="pl-v">SNPFilter</span> <span class="pl-k">=</span> MyFilter())</pre></div>
<p>And if you ever get lost, there's an online <strong>help()</strong> function for each object type:</p>
<div class="highlight highlight-source-python"><pre><span class="pl-k">from</span> pyGeno.Genome <span class="pl-k">import</span> <span class="pl-k">*</span>

<span class="pl-c1">print</span> Exon.help()</pre></div>
<p>Should output:</p>
<pre>Available fields for Exon: CDS_start, end, chromosome, CDS_length, frame, number, CDS_end, start, genome, length, protein, gene, transcript, id, strand
</pre>
<a name="user-content-creating-a-personalized-genome"/>
<h2><a id="user-content-creating-a-personalized-genome" class="anchor" href="#creating-a-personalized-genome" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" role="img" version="1.1" viewbox="0 0 16 16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"/></svg></a>Creating a Personalized Genome:</h2>
<p>Personalized Genomes are a powerful feature that allow you to work on the specific genomes and proteomes of your patients. You can even mix several SNP sets together.</p>
<div class="highlight highlight-source-python"><pre><span class="pl-k">from</span> pyGeno.Genome <span class="pl-k">import</span> Genome
<span class="pl-c">#the name of the snp set is defined inside the datawrap's manifest.ini file</span>
dummy <span class="pl-k">=</span> Genome(<span class="pl-v">name</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span>GRCh37.75<span class="pl-pds">'</span></span>, <span class="pl-v">SNPs</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span>dummySRY<span class="pl-pds">'</span></span>)
<span class="pl-c">#you can also define a filter (ex: a quality filter) for the SNPs</span>
dummy <span class="pl-k">=</span> Genome(<span class="pl-v">name</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span>GRCh37.75<span class="pl-pds">'</span></span>, <span class="pl-v">SNPs</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span>dummySRY<span class="pl-pds">'</span></span>, <span class="pl-v">SNPFilter</span> <span class="pl-k">=</span> myFilter())
<span class="pl-c">#and even mix several snp sets</span>
dummy <span class="pl-k">=</span> Genome(<span class="pl-v">name</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span>GRCh37.75<span class="pl-pds">'</span></span>, <span class="pl-v">SNPs</span> <span class="pl-k">=</span> [<span class="pl-s"><span class="pl-pds">'</span>dummySRY<span class="pl-pds">'</span></span>, <span class="pl-s"><span class="pl-pds">'</span>anotherSet<span class="pl-pds">'</span></span>], <span class="pl-v">SNPFilter</span> <span class="pl-k">=</span> myFilter())</pre></div>
<a name="user-content-filtering-snps"/>
<h2><a id="user-content-filtering-snps" class="anchor" href="#filtering-snps" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" role="img" version="1.1" viewbox="0 0 16 16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"/></svg></a>Filtering SNPs:</h2>
<p>pyGeno allows you to select the Polymorphisms that end up into the final sequences. It supports SNPs, Inserts and Deletions.</p>
<div class="highlight highlight-source-python"><pre><span class="pl-k">from</span> pyGeno.SNPFiltering <span class="pl-k">import</span> SNPFilter, SequenceSNP

<span class="pl-k">class</span> <span class="pl-en">QMax_gt_filter</span>(<span class="pl-e">SNPFilter</span>) :

        <span class="pl-k">def</span> <span class="pl-c1">__init__</span>(<span class="pl-smi"><span class="pl-smi">self</span></span>, <span class="pl-smi">threshold</span>) :
                <span class="pl-v">self</span>.threshold <span class="pl-k">=</span> threshold

        <span class="pl-c">#Here SNPs is a dictionary: SNPSet Name =&gt; polymorphism</span>
        <span class="pl-c">#This filter ignores deletions and insertions and</span>
        <span class="pl-c">#but applis all SNPs</span>
        <span class="pl-k">def</span> <span class="pl-c1">filter</span>(<span class="pl-smi"><span class="pl-smi">self</span></span>, <span class="pl-smi">chromosome</span>, <span class="pl-k">**</span><span class="pl-smi">SNPs</span>) :
                sources <span class="pl-k">=</span> {}
                alleles <span class="pl-k">=</span> []
                <span class="pl-k">for</span> snpSet, snp <span class="pl-k">in</span> SNPs.iteritems() :
                        pos <span class="pl-k">=</span> snp.start
                        <span class="pl-k">if</span> snp.alt[<span class="pl-c1">0</span>] <span class="pl-k">==</span> <span class="pl-s"><span class="pl-pds">'</span>-<span class="pl-pds">'</span></span> :
                                <span class="pl-k">pass</span>
                        <span class="pl-k">elif</span> snp.ref[<span class="pl-c1">0</span>] <span class="pl-k">==</span> <span class="pl-s"><span class="pl-pds">'</span>-<span class="pl-pds">'</span></span> :
                                <span class="pl-k">pass</span>
                        <span class="pl-k">else</span> :
                                sources[snpSet] <span class="pl-k">=</span> snp
                                alleles.append(snp.alt) <span class="pl-c">#if not an indel append the polymorphism</span>

                <span class="pl-c">#appends the refence allele to the lot</span>
                refAllele <span class="pl-k">=</span> chromosome.refSequence[pos]
                alleles.append(refAllele)
                sources[<span class="pl-s"><span class="pl-pds">'</span>ref<span class="pl-pds">'</span></span>] <span class="pl-k">=</span> refAllele

                <span class="pl-c">#optional we keep a record of the polymorphisms that were used during the process</span>
                <span class="pl-k">return</span> SequenceSNP(alleles, <span class="pl-v">sources</span> <span class="pl-k">=</span> sources)</pre></div>
<p>The filter function can also be made more specific by using arguments that have the same names as the SNPSets</p>
<div class="highlight highlight-source-python"><pre><span class="pl-k">def</span> <span class="pl-c1">filter</span>(<span class="pl-smi"><span class="pl-smi">self</span></span>, <span class="pl-smi">chromosome</span>, <span class="pl-smi">dummySRY</span> <span class="pl-k">=</span> <span class="pl-c1">None</span>) :
        <span class="pl-k">if</span> dummySRY.Qmax_gt <span class="pl-k">&gt;</span> <span class="pl-v">self</span>.threshold :
                <span class="pl-c">#other possibilities of return are SequenceInsert(&lt;bases&gt;), SequenceDelete(&lt;length&gt;)</span>
                <span class="pl-k">return</span> SequenceSNP(dummySRY.alt)
        <span class="pl-k">return</span> <span class="pl-c1">None</span> <span class="pl-c">#None means keep the reference allele</span></pre></div>
<p>To apply the filter simply specify if while loading the genome.</p>
<div class="highlight highlight-source-python"><pre>persGenome <span class="pl-k">=</span> Genome(<span class="pl-v">name</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span>GRCh37.75_Y-Only<span class="pl-pds">'</span></span>, <span class="pl-v">SNPs</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span>dummySRY<span class="pl-pds">'</span></span>, <span class="pl-v">SNPFilter</span> <span class="pl-k">=</span> QMax_gt_filter(<span class="pl-c1">10</span>))</pre></div>
<p>To include several SNPSets use a list.</p>
<div class="highlight highlight-source-python"><pre>persGenome <span class="pl-k">=</span> Genome(<span class="pl-v">name</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span>GRCh37.75_Y-Only<span class="pl-pds">'</span></span>, <span class="pl-v">SNPs</span> <span class="pl-k">=</span> [<span class="pl-s"><span class="pl-pds">'</span>ARN_P1<span class="pl-pds">'</span></span>, <span class="pl-s"><span class="pl-pds">'</span>ARN_P2<span class="pl-pds">'</span></span>], <span class="pl-v">SNPFilter</span> <span class="pl-k">=</span> myFilter())</pre></div>
<a name="user-content-getting-an-arbitrary-sequence"/>
<h2><a id="user-content-getting-an-arbitrary-sequence" class="anchor" href="#getting-an-arbitrary-sequence" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" role="img" version="1.1" viewbox="0 0 16 16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"/></svg></a>Getting an arbitrary sequence:</h2>
<p>You can ask for any sequence of any chromosome:</p>
<div class="highlight highlight-source-python"><pre>chr12 <span class="pl-k">=</span> myGenome.get(Chromosome, <span class="pl-v">number</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">"</span>12<span class="pl-pds">"</span></span>)[<span class="pl-c1">0</span>]
<span class="pl-c1">print</span> chr12.sequence[x1:x2]
<span class="pl-c"># for the reference sequence</span>
<span class="pl-c1">print</span> chr12.refSequence[x1:x2]</pre></div>
<a name="user-content-batteries-included-bootstraping"/>
<h2><a id="user-content-batteries-included-bootstraping" class="anchor" href="#batteries-included-bootstraping" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" role="img" version="1.1" viewbox="0 0 16 16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"/></svg></a>Batteries included (bootstraping):</h2>
<p>pyGeno's database is populated by importing datawraps.
pyGeno comes with a few data wraps, to get the list you can use:</p>
<div class="highlight highlight-source-python"><pre><span class="pl-k">import</span> pyGeno.bootstrap <span class="pl-k">as</span> <span class="pl-c1">B</span>
<span class="pl-c1">B</span>.printDatawraps()</pre></div>
<pre>Available datawraps for boostraping

SNPs
~~~~|
    |~~~:&gt; Human_agnostic.dummySRY.tar.gz
    |~~~:&gt; Human.dummySRY_casava.tar.gz
    |~~~:&gt; dbSNP142_human_common_all.tar.gz


Genomes
~~~~~~~|
       |~~~:&gt; Human.GRCh37.75.tar.gz
       |~~~:&gt; Human.GRCh37.75_Y-Only.tar.gz
       |~~~:&gt; Human.GRCh38.78.tar.gz
       |~~~:&gt; Mouse.GRCm38.78.tar.gz
</pre>
<p>To get a list of remote datawraps that pyGeno can download for you, do:</p>

<p>Importing whole genomes is a demanding process that take more than an hour and requires (according to tests)
at least 3GB of memory. Depending on your configuration, more might be required.</p>
<p>That being said importating a data wrap is a one time operation and once the importation is complete the datawrap
can be discarded without consequences.</p>
<p>The bootstrap module also has some handy functions for importing built-in packages.</p>
<p>Some of them just for playing around with pyGeno (<strong>Fast importation</strong> and <strong>Small memory requirements</strong>):</p>
<div class="highlight highlight-source-python"><pre><span class="pl-k">import</span> pyGeno.bootstrap <span class="pl-k">as</span> <span class="pl-c1">B</span>

<span class="pl-c">#Imports only the Y chromosome from the human reference genome GRCh37.75</span>
<span class="pl-c">#Very fast, requires even less memory. No download required.</span>
<span class="pl-c1">B</span>.importGenome(<span class="pl-s"><span class="pl-pds">"</span>Human.GRCh37.75_Y-Only.tar.gz<span class="pl-pds">"</span></span>)

<span class="pl-c">#A dummy datawrap for humans SNPs and Indels in pyGeno's AgnosticSNP  format.</span>
<span class="pl-c"># This one has one SNP at the begining of the gene SRY</span>
<span class="pl-c1">B</span>.importSNPs(<span class="pl-s"><span class="pl-pds">"</span>Human.dummySRY_casava.tar.gz<span class="pl-pds">"</span></span>)</pre></div>
<p>And for more <strong>Serious Work</strong>, the whole reference genome.</p>
<div class="highlight highlight-source-python"><pre><span class="pl-c">#Downloads the whole genome (205MB, sequences + annotations), may take an hour or more.</span>
<span class="pl-c1">B</span>.importGenome(<span class="pl-s"><span class="pl-pds">"</span>Human.GRCh38.78.tar.gz<span class="pl-pds">"</span></span>)</pre></div>
<a name="user-content-importing-a-custom-datawrap"/>
<h2><a id="user-content-importing-a-custom-datawrap" class="anchor" href="#importing-a-custom-datawrap" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" role="img" version="1.1" viewbox="0 0 16 16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"/></svg></a>Importing a custom datawrap:</h2>
<div class="highlight highlight-source-python"><pre><span class="pl-k">from</span> pyGeno.importation.Genomes <span class="pl-k">import</span> <span class="pl-k">*</span>
importGenome(<span class="pl-s"><span class="pl-pds">'</span>GRCh37.75.tar.gz<span class="pl-pds">'</span></span>)</pre></div>
<p>To import a patient's specific polymorphisms</p>
<div class="highlight highlight-source-python"><pre><span class="pl-k">from</span> pyGeno.importation.SNPs <span class="pl-k">import</span> <span class="pl-k">*</span>
importSNPs(<span class="pl-s"><span class="pl-pds">'</span>patient1.tar.gz<span class="pl-pds">'</span></span>)</pre></div>
<p>For a list of available datawraps available for download, please have a look <a href="http://pygeno.iric.ca/datawraps.html">here</a>.</p>
<p>You can easily make your own datawraps with any tar.gz compressor.
For more details on how datawraps are made you can check <a href="https://github.com/tariqdaouda/pyGeno/wiki/How-to-create-a-pyGeno-friendly-package-to-import-your-data%3F">wiki</a> or have a look inside the folder bootstrap_data.</p>
<a name="user-content-instanciating-a-genome"/>
<h2><a id="user-content-instanciating-a-genome" class="anchor" href="#instanciating-a-genome" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" role="img" version="1.1" viewbox="0 0 16 16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"/></svg></a>Instanciating a genome:</h2>
<div class="highlight highlight-source-python"><pre><span class="pl-k">from</span> pyGeno.Genome <span class="pl-k">import</span> Genome
<span class="pl-c">#the name of the genome is defined inside the package's manifest.ini file</span>
ref <span class="pl-k">=</span> Genome(<span class="pl-v">name</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span>GRCh37.75<span class="pl-pds">'</span></span>)</pre></div>
<a name="user-content-printing-all-the-proteins-of-a-gene"/>
<h2><a id="user-content-printing-all-the-proteins-of-a-gene" class="anchor" href="#printing-all-the-proteins-of-a-gene" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" role="img" version="1.1" viewbox="0 0 16 16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"/></svg></a>Printing all the proteins of a gene:</h2>
<div class="highlight highlight-source-python"><pre><span class="pl-k">from</span> pyGeno.Genome <span class="pl-k">import</span> Genome
<span class="pl-k">from</span> pyGeno.Gene <span class="pl-k">import</span> Gene
<span class="pl-k">from</span> pyGeno.Protein <span class="pl-k">import</span> Protein</pre></div>
<p>Or simply:</p>
<div class="highlight highlight-source-python"><pre><span class="pl-k">from</span> pyGeno.Genome <span class="pl-k">import</span> <span class="pl-k">*</span></pre></div>
<p>then:</p>
<div class="highlight highlight-source-python"><pre>ref <span class="pl-k">=</span> Genome(<span class="pl-v">name</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span>GRCh37.75<span class="pl-pds">'</span></span>)
<span class="pl-c">#get returns a list of elements</span>
gene <span class="pl-k">=</span> ref.get(Gene, <span class="pl-v">name</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span>TPST2<span class="pl-pds">'</span></span>)[<span class="pl-c1">0</span>]
<span class="pl-k">for</span> prot <span class="pl-k">in</span> gene.get(Protein) :
      <span class="pl-c1">print</span> prot.sequence</pre></div>
<a name="user-content-making-queries-get-vs-iterget"/>
<h2><a id="user-content-making-queries-get-vs-iterget" class="anchor" href="#making-queries-get-vs-iterget" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" role="img" version="1.1" viewbox="0 0 16 16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"/></svg></a>Making queries, get() Vs iterGet():</h2>
<p>iterGet is a faster version of get that returns an iterator instead of a list.</p>
<a name="user-content-making-queries-syntax"/>
<h2><a id="user-content-making-queries-syntax" class="anchor" href="#making-queries-syntax" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" role="img" version="1.1" viewbox="0 0 16 16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"/></svg></a>Making queries, syntax:</h2>
<p>pyGeno's get function uses the expressivity of rabaDB.</p>
<p>These are all possible query formats:</p>
<div class="highlight highlight-source-python"><pre>ref.get(Gene, <span class="pl-v">name</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">"</span>SRY<span class="pl-pds">"</span></span>)
ref.get(Gene, { <span class="pl-s"><span class="pl-pds">"</span>name like<span class="pl-pds">"</span></span> : <span class="pl-s"><span class="pl-pds">"</span>HLA<span class="pl-pds">"</span></span>})
chr12.get(Exon, { <span class="pl-s"><span class="pl-pds">"</span>start &gt;=<span class="pl-pds">"</span></span> : <span class="pl-c1">12000</span>, <span class="pl-s"><span class="pl-pds">"</span>end &lt;<span class="pl-pds">"</span></span> : <span class="pl-c1">12300</span> })
ref.get(Transcript, { <span class="pl-s"><span class="pl-pds">"</span>gene.name<span class="pl-pds">"</span></span> : <span class="pl-s"><span class="pl-pds">'</span>SRY<span class="pl-pds">'</span></span> })</pre></div>
<a name="user-content-creating-indexes-to-speed-up-queries"/>
<h2><a id="user-content-creating-indexes-to-speed-up-queries" class="anchor" href="#creating-indexes-to-speed-up-queries" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" role="img" version="1.1" viewbox="0 0 16 16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"/></svg></a>Creating indexes to speed up queries:</h2>
<div class="highlight highlight-source-python"><pre><span class="pl-k">from</span> pyGeno.Gene <span class="pl-k">import</span> Gene
<span class="pl-c">#creating an index on gene names if it does not already exist</span>
Gene.ensureGlobalIndex(<span class="pl-s"><span class="pl-pds">'</span>name<span class="pl-pds">'</span></span>)
<span class="pl-c">#removing the index</span>
Gene.dropIndex(<span class="pl-s"><span class="pl-pds">'</span>name<span class="pl-pds">'</span></span>)</pre></div>
<a name="user-content-find-in-sequences"/>
<h2><a id="user-content-find-in-sequences" class="anchor" href="#find-in-sequences" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" role="img" version="1.1" viewbox="0 0 16 16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"/></svg></a>Find in sequences:</h2>
<p>Internally pyGeno uses a binary representation for nucleotides and amino acids to deal with polymorphisms.
For example,both "AGC" and "ATG" will match the following sequence "...AT/GCCG...".</p>
<div class="highlight highlight-source-python"><pre><span class="pl-c">#returns the position of the first occurence</span>
transcript.find(<span class="pl-s"><span class="pl-pds">"</span>AT/GCCG<span class="pl-pds">"</span></span>)
<span class="pl-c">#returns the positions of all occurences</span>
transcript.findAll(<span class="pl-s"><span class="pl-pds">"</span>AT/GCCG<span class="pl-pds">"</span></span>)

<span class="pl-c">#similarly, you can also do</span>
transcript.findIncDNA(<span class="pl-s"><span class="pl-pds">"</span>AT/GCCG<span class="pl-pds">"</span></span>)
transcript.findAllIncDNA(<span class="pl-s"><span class="pl-pds">"</span>AT/GCCG<span class="pl-pds">"</span></span>)
transcript.findInUTR3(<span class="pl-s"><span class="pl-pds">"</span>AT/GCCG<span class="pl-pds">"</span></span>)
transcript.findAllInUTR3(<span class="pl-s"><span class="pl-pds">"</span>AT/GCCG<span class="pl-pds">"</span></span>)
transcript.findInUTR5(<span class="pl-s"><span class="pl-pds">"</span>AT/GCCG<span class="pl-pds">"</span></span>)
transcript.findAllInUTR5(<span class="pl-s"><span class="pl-pds">"</span>AT/GCCG<span class="pl-pds">"</span></span>)

<span class="pl-c">#same for proteins</span>
protein.find(<span class="pl-s"><span class="pl-pds">"</span>DEV/RDEM<span class="pl-pds">"</span></span>)
protein.findAll(<span class="pl-s"><span class="pl-pds">"</span>DEV/RDEM<span class="pl-pds">"</span></span>)

<span class="pl-c">#and for exons</span>
exon.find(<span class="pl-s"><span class="pl-pds">"</span>AT/GCCG<span class="pl-pds">"</span></span>)
exon.findAll(<span class="pl-s"><span class="pl-pds">"</span>AT/GCCG<span class="pl-pds">"</span></span>)
exon.findInCDS(<span class="pl-s"><span class="pl-pds">"</span>AT/GCCG<span class="pl-pds">"</span></span>)
exon.findAllInCDS(<span class="pl-s"><span class="pl-pds">"</span>AT/GCCG<span class="pl-pds">"</span></span>)
<span class="pl-c">#...</span></pre></div>
<a name="user-content-progress-bar"/>
<h2><a id="user-content-progress-bar" class="anchor" href="#progress-bar" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" role="img" version="1.1" viewbox="0 0 16 16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"/></svg></a>Progress Bar:</h2>
<div class="highlight highlight-source-python"><pre><span class="pl-k">from</span> pyGeno.tools.ProgressBar <span class="pl-k">import</span> ProgressBar
pg <span class="pl-k">=</span> ProgressBar(<span class="pl-v">nbEpochs</span> <span class="pl-k">=</span> <span class="pl-c1">155</span>)
<span class="pl-k">for</span> i <span class="pl-k">in</span> <span class="pl-c1">range</span>(<span class="pl-c1">155</span>) :
      pg.update(<span class="pl-v">label</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span><span class="pl-c1">%d</span><span class="pl-pds">'</span></span> <span class="pl-k">%</span>i) <span class="pl-c"># or simply p.update()</span>
pg.close()</pre></div>

</article>
  </div></body></html>