<html><body><div><div class="post">
  <h1 class="post-title">Exploring the Genome with Ensembl and Python</h1>
  <span class="post-date">04 Feb 2015</span>
  <p>The <a href="http://www.genome.gov/12011238">sequencing of the human genome</a> took 13 years to complete, cost $3 billion dollars, was lauded as “<a href="http://partners.nytimes.com/library/national/science/062700sci-genome.html"><em>a pinnacle of human self-knowledge</em></a>”, and even inspired <a href="http://www.thehumangenome.co.uk/THE_HUMAN_GENOME/Home.html">a book of poetry</a>. The actual result of this project is a small collection of very long sequences (one for each chromosome), varying in length from tens to hundreds of millions of nucleotides (<a href="http://en.wikipedia.org/wiki/Adenine">A</a>’s, <a href="http://en.wikipedia.org/wiki/Cytosine">C</a>’s, <a href="http://en.wikipedia.org/wiki/Thymine">T</a>’s, and <a href="http://en.wikipedia.org/wiki/Guanine">G</a>’s). For comparison, the King James Bible contains only about 3 million letters, which would make for a very diminutive chromosome.</p>

<p>What’s in this colossal nucleic tome? The sequence of the human genome only hints at its structure and organization.
Where are the <a href="http://www.quora.com/What-is-a-gene-in-modern-biology">genes</a>? Which parts are <a href="http://en.wikipedia.org/wiki/Primary_transcript">transcribed</a> and how are they <a href="http://bitesizebio.com/10148/what-is-alternative-splicing-and-why-is-it-important/">spliced</a>? We need <a href="http://en.wikipedia.org/wiki/Genome_project#Genome_annotation">annotations</a> on top of the sequence to tell us how the genome gets read.</p>

<h2 id="what-is-ensembl">What is Ensembl?</h2>

<p>Realizing that sequencing the genome is only half the battle, in 1999 the <a href="http://www.embl.de/">European Molecular Biology Laboratory</a> partnered with the <a href="https://www.sanger.ac.uk/">Sanger Institute</a> to make an annotation database called <a href="http://www.ensembl.org">Ensembl</a>.
Ensembl catalogues where all the suspected active elements in a genome reside and to what degree they have been validated experimentally.</p>

<p>Many different kinds of information are gathered together within Ensembl but the most fundamental annotations are:</p>

<ul>
  <li>
    <p><strong>Genes</strong>: Transcriptionally active regions of the genome which may give rise to several related <a href="http://en.wikipedia.org/wiki/Protein">proteins</a> or <a href="http://en.wikipedia.org/wiki/Non-coding_RNA">non-coding RNA</a>s. A gene is characterized by a range of positions on a particular chromosome and can contain multiple transcripts.</p>
  </li>
  <li>
    <p><strong>Transcripts</strong>: A subset of a gene which gets copied into RNA (by a process called <a href="http://en.wikipedia.org/wiki/Transcription_(genetics)">transcription</a>). The subset of transcripts which are destined to become <a href="http://en.wikipedia.org/wiki/Messenger_RNA">messenger RNAs</a> (used as templates for constructing proteins) are also annotated with both their <a href="http://en.wikipedia.org/wiki/RNA_splicing">splice boundaries</a> and the specific <a href="http://www.broadinstitute.org/blog/what-orf">open reading frame</a> that gets used for <a href="http://www.hhmi.org/biointeractive/translation-advanced-detail">translation</a>.</p>
  </li>
</ul>

<div class="image-container">
<img src="/images/gene_to_protein.png" alt="Gene, transcript, mRNA, protein"/>
<div class="image-caption"><p>
    Gene, transcript, mRNA, protein
    (from
    </p><a href="http://en.wikipedia.org/wiki/Regulation_of_gene_expression#mediaviewer/File:Gene_expression_control.png">wikipedia</a><p>)
</p></div>
</div>

<h2 id="how-to-access-ensembl">How to access Ensembl</h2>

<p>Ensembl provides a very detailed <a href="http://www.ensembl.org/Homo_sapiens/Info/Index">web interface</a>, through which you can inspect all the information associated with a particular <a href="http://useast.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000139618;r=13:32315474-32400266">gene</a>, <a href="http://useast.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000139618;r=13:32315474-32400266;t=ENST00000380152">transcript</a>, <a href="http://useast.ensembl.org/Homo_sapiens/Variation/Phenotype?ph=5996;r=3:52686789-52687789;v=rs11177;vdb=variation;vf=22329">genetic variant</a>, or even the genetic locations associated with a  <a href="http://useast.ensembl.org/Homo_sapiens/Phenotype/Locations?ph=5996">disease</a>. If you get tired of clicking around the Ensembl website, it’s also possible to issue programmatic queries via the <a href="http://rest.ensembl.org/">Ensembl REST API</a>.</p>

<p>Though convenient, using a web API restricts how many distinct queries you can issue in a small amount of time. If you want to deal with thousands of genomic entities or locations, then it would be preferable to use a local copy of Ensembl’s data. Unfortunately, <a href="http://useast.ensembl.org/info/docs/webcode/mirror/install/index.html?redirect=no">local installation</a> of Ensembl is complicated and requires an up-front decision about what subset of the data you’d like to use. Furthermore, while there is a convenient <a href="http://useast.ensembl.org/info/docs/api/index.html">Perl API</a> for local queries, Ensembl lacks an equivalent for Python.</p>

<h2 id="introducing-pyensembl">Introducing PyEnsembl</h2>

<p>To facilitate fast/local querying of Ensembl annotations from within Python we wrote <a href="https://github.com/hammerlab/pyensembl">PyEnsembl</a>.
PyEnsembl lazily downloads annotation datasets from the <a href="http://www.ensembl.org/info/data/ftp/index.html">Ensembl FTP server</a>, which it then uses to populate a local <a href="https://docs.python.org/2/library/sqlite3.html">sqlite</a> database. You don’t have to interact with the database or its schema directly. Instead, you can use a particular snapshot of the Ensembl database by creating a <code class="highlighter-rouge">pyensembl.EnsemblRelease</code> object:</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="kn">import</span> <span class="nn">pyensembl</span>
<span class="n">ensembl</span> <span class="o">=</span> <span class="n">pyensembl</span><span class="o">.</span><span class="n">EnsemblRelease</span><span class="p">()</span>
</code></pre>
</div>

<p><code class="highlighter-rouge">EnsemblRelease</code> constructs queries for you through a variety of helper methods (such as <code class="highlighter-rouge">locations_of_gene_name</code>), some of which return structured objects that correspond to annotated entities (such as <code class="highlighter-rouge">Gene</code>, <code class="highlighter-rouge">Transcript</code>, and <code class="highlighter-rouge">Exon</code>).</p>

<p>By default, PyEnsembl will use the latest Ensembl release (currently 78). If, however, you want to switch to an older Ensembl release you can specify it as an argument to the <code class="highlighter-rouge">EnsemblRelease</code> constructor:</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">ensembl</span> <span class="o">=</span> <span class="n">pyensembl</span><span class="o">.</span><span class="n">EnsemblRelease</span><span class="p">(</span><span class="n">release</span><span class="o">=</span><span class="mi">75</span><span class="p">)</span>
</code></pre>
</div>

<p>Note: the first time you use a particular release, there will be a delay of several minutes while the library downloads and parses Ensembl data files.</p>

<h2 id="example-what-gene-is-at-a-particular-position">Example: What gene is at a particular position?</h2>

<p>We can use the helper method <code class="highlighter-rouge">EnsemblRelease.genes_at_locus</code> to determine
what genes (if any) overlap some region of a chromosome.</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">genes</span> <span class="o">=</span> <span class="n">ensembl</span><span class="o">.</span><span class="n">genes_at_locus</span><span class="p">(</span><span class="n">contig</span><span class="o">=</span><span class="s">"1"</span><span class="p">,</span> <span class="n">position</span><span class="o">=</span><span class="mi">1000000</span><span class="p">)</span>
</code></pre>
</div>

<p>The parameter <code class="highlighter-rouge">contig</code> refers to which contiguous region of DNA we’re interested in (typically a chromosome). The result contained in <code class="highlighter-rouge">genes</code> is <code class="highlighter-rouge">[Gene(id=ENSG00000188290, name=HES4)]</code>. This means that the transcription factor <a href="http://www.genecards.org/cgi-bin/carddisp.pl?gene=HES4">HES4</a> overlaps the millionth nucleotide of chromosome 1. What are the exact boundaries of this gene? Let’s pull it out
of the list <code class="highlighter-rouge">genes</code> and inspect its location properties.</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">hes4</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">hes4_location</span> <span class="o">=</span> <span class="p">(</span><span class="n">hes4</span><span class="o">.</span><span class="n">contig</span><span class="p">,</span> <span class="n">hes4</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">hes4</span><span class="o">.</span><span class="n">end</span><span class="p">)</span>
</code></pre>
</div>

<p>The value of <code class="highlighter-rouge">hes4_location</code> will be <code class="highlighter-rouge">('1', 998962, 1000172)</code>, meaning that this whole gene spans only 1,210 nucleotides of chromosome 1.</p>

<p>Does HES4 have any known transcripts? By looking at <code class="highlighter-rouge">hes4.transcripts</code> you’ll see that it has 4 annotated transcripts: HES4-001, HES4-002, HES4-003, and HES4-004. You can get the <a href="http://humangenes.org/cdna-complementary-dna">cDNA</a> sequence of any transcript by accessing its <code class="highlighter-rouge">sequence</code> field. You only can also look at only the protein coding portion of a sequence by accessing the <code class="highlighter-rouge">coding_sequence</code> field of a transcript.</p>

<h2 id="example-how-many-genes-are-there-in-total">Example: How many genes are there in total?</h2>

<p>To determine the total number of known genes, let’s start by getting the IDs of all the genes in Ensembl.</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">gene_ids</span> <span class="o">=</span> <span class="n">ensembl</span><span class="o">.</span><span class="n">gene_ids</span><span class="p">()</span>
</code></pre>
</div>

<p>The entries of the list <code class="highlighter-rouge">gene_ids</code> are all distinct string identifiers (e.g. “ENSG00000238009”).
If you count the number of IDs, you’ll get a shockingly large number: 64,253! Aren’t there supposed to be <a href="https://www.sciencenews.org/article/more-chicken-fewer-grape">around 22,000 genes</a>? Where did all the extra identifiers come from?</p>

<p>The Ensembl definition of a gene (a transcriptionally active region) is more liberal than the common usage (which typically assumes that each gene produces some protein). To count how many <em>protein coding genes</em> are annotated in Ensembl, we’ll have to look at the <a href="http://www.gencodegenes.org/gencode_biotypes.html">biotype</a> associated with each gene.</p>

<p>To get these <em>biotypes</em>, let’s first construct a list of <code class="highlighter-rouge">Gene</code> objects for each ID in <code class="highlighter-rouge">gene_ids</code>.</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">ensembl</span><span class="o">.</span><span class="n">gene_by_id</span><span class="p">(</span><span class="n">gene_id</span><span class="p">)</span> <span class="k">for</span> <span class="n">gene_id</span> <span class="ow">in</span> <span class="n">gene_ids</span><span class="p">]</span>
</code></pre>
</div>

<p>We still have a list of 64,253 elements, but now the entries look like <code class="highlighter-rouge">Gene(id=ENSG00000238009, name=RP11-34P13.7)</code>. Each of these <code class="highlighter-rouge">Gene</code> objects has a <code class="highlighter-rouge">biotype</code> field, which can take on a daunting zoo of values such as “lincRNA”, “rRNA”, or “pseudogene”. For our purposes, the only value we care about is “protein_coding”.</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">coding_genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span> <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">biotype</span> <span class="o">==</span> <span class="s">'protein_coding'</span><span class="p">]</span>
</code></pre>
</div>

<p>The length of <code class="highlighter-rouge">coding_genes</code> is much more in line with our expectations: 21,983.</p>

<h2 id="limitations-and-roadmap">Limitations and Roadmap</h2>

<p>Hopefully the two examples above give you a flavor of PyEnsembl’s expressiveness for working with genome annotations. It’s worth pointing out that PyEnsembl is a very new library and still missing significant functionality.</p>

<ul>
  <li>PyEnsembl currently only supports human annotations. Future releases will support other species by specifying a <code class="highlighter-rouge">species</code> argument to <code class="highlighter-rouge">EnsemblRelease</code> (e.g. <code class="highlighter-rouge">EnsemblRelease(release=78, species="mouse")</code>).</li>
  <li>The first time you use PyEnsembl with a new release, the library will spend several minutes parsing its data into a sqlite table. This delay could be cut down significantly by rewriting the data parsing logic in <a href="http://cython.org/">Cython</a>.</li>
  <li><code class="highlighter-rouge">Transcript</code> objects currently only have access to their spliced sequence but not to the sequences of their <a href="http://en.wikipedia.org/wiki/Intron">introns</a>.</li>
  <li>PyEnsembl currently only exposes core annotations (those found in Ensembl’s <a href="http://www.ensembl.org/info/website/upload/gff.html">GTF</a> files), but lacks many secondary annotations (such as <a href="http://appris.bioinfo.cnio.es/">APPRIS</a>).</li>
</ul>

<p>You can install PyEnsembl using <code class="highlighter-rouge">pip</code> by running <code class="highlighter-rouge">pip install pyensembl</code>. If you encounter any problems with PyEnsembl or have suggestions for improving the library, please file an issue <a href="https://github.com/hammerlab/pyensembl">on GitHub</a>.</p>

</div>

      </div></body></html>