<html><body><div><div class="section">
              <h1>GPy 0.9.1</h1>

              


<p>The Gaussian Process Toolbox</p>

<p>
   <a href="http://pythonhosted.org/GPy/">Package Documentation</a>
</p>






<p>The Gaussian processes framework in Python.</p>

<div id="updated-structure">
<h2>Updated Structure</h2>
<p>We have pulled the core parameterization out of GPy. It is a package
called <a href="https://github.com/sods/paramz" rel="nofollow">paramz</a> and is the pure
gradient based model optimization.</p>
<p>If you installed GPy with pip, just upgrade the package using:</p>
<pre>$ pip install --upgrade GPy
</pre>
<p>If you have the developmental version of GPy (using the develop or -e
option) just install the dependencies by running</p>
<pre>$ python setup.py develop
</pre>
<p>again, in the GPy installation folder.</p>
<p>A warning: This usually works, but sometimes <tt>distutils/setuptools</tt>
opens a whole can of worms here, specially when compiled extensions are
involved. If that is the case, it is best to clean the repo and
reinstall.</p>
</div>
<div id="continuous-integration">
<h2>Continuous integration</h2>
<table>
<colgroup>
<col/>
<col/>
<col/>
<col/>
</colgroup>
<thead>
<tr><th> </th>
<th>Travis-CI</th>
<th>Codecov</th>
<th>RTFD</th>
</tr>
</thead>
<tbody>
<tr><td><strong>master:</strong></td>
<td><a href="https://travis-ci.org/SheffieldML/GPy" rel="nofollow"><img src="https://travis-ci.org/SheffieldML/GPy.svg?branch=master"/></a></td>
<td><a href="http://codecov.io/github/SheffieldML/GPy?branch=master" rel="nofollow"><img src="http://codecov.io/github/SheffieldML/GPy/coverage.svg?branch=master"/></a></td>
<td><a href="http://gpy.readthedocs.org/en/master/" rel="nofollow"><img src="https://readthedocs.org/projects/gpy/badge/?version=master"/></a></td>
</tr>
<tr><td><strong>devel:</strong></td>
<td><a href="https://travis-ci.org/SheffieldML/GPy" rel="nofollow"><img src="https://travis-ci.org/SheffieldML/GPy.svg?branch=devel"/></a></td>
<td><a href="http://codecov.io/github/SheffieldML/GPy?branch=devel" rel="nofollow"><img src="http://codecov.io/github/SheffieldML/GPy/coverage.svg?branch=devel"/></a></td>
<td><a href="http://gpy.readthedocs.org/en/devel/" rel="nofollow"><img src="https://readthedocs.org/projects/gpy/badge/?version=devel"/></a></td>
</tr>
</tbody>
</table>
</div>

<div id="citation">
<h2>Citation</h2>
<pre>@Misc{gpy2014,
  author =   {{The GPy authors}},
  title =    {{GPy}: A Gaussian process framework in python},
  howpublished = {\url{http://github.com/SheffieldML/GPy}},
  year = {2012--2015}
}
</pre>
<div id="pronounciation">
<h3>Pronounciation:</h3>
<p>We like to pronounce it ‘g-pie’.</p>
</div>
</div>
<div id="getting-started-installing-with-pip">
<h2>Getting started: installing with pip</h2>
<p>We are now requiring the newest version (0.16) of
<a href="http://www.scipy.org/" rel="nofollow">scipy</a> and thus, we strongly recommend using
the <a href="http://continuum.io/downloads" rel="nofollow">anaconda python distribution</a>.
With anaconda you can install GPy by the following:</p>
<pre>conda update scipy
pip install gpy
</pre>
<p>We’ve also had luck with <a href="http://www.enthought.com" rel="nofollow">enthought</a>.
Install scipy 0.16 (or later) and then pip install GPy:</p>
<pre>pip install gpy
</pre>
<p>If you’d like to install from source, or want to contribute to the
project (i.e. by sending pull requests via github), read on.</p>
<div id="troubleshooting-installation-problems">
<h3>Troubleshooting installation problems</h3>
<p>If you’re having trouble installing GPy via <tt>pip install GPy</tt> here is
a probable solution:</p>
<pre>git clone https://github.com/SheffieldML/GPy.git
cd GPy
git checkout devel
python setup.py build_ext --inplace
nosetests GPy/testing
</pre>
</div>

</div>
<div id="running-unit-tests">
<h2>Running unit tests:</h2>
<p>Ensure nose is installed via pip:</p>
<pre>pip install nose
</pre>
<p>Run nosetests from the root directory of the repository:</p>
<pre>nosetests -v GPy/testing
</pre>
<p>or from within IPython</p>
<pre>import GPy; GPy.tests()
</pre>
<p>or using setuptools</p>
<pre>python setup.py test
</pre>
</div>
<div id="ubuntu-hackers">
<h2>Ubuntu hackers</h2>
<blockquote>
Note: Right now the Ubuntu package index does not include scipy
0.16.0, and thus, cannot be used for GPy. We hope this gets fixed
soon.</blockquote>
<p>For the most part, the developers are using ubuntu. To install the
required packages:</p>
<pre>sudo apt-get install python-numpy python-scipy python-matplotlib
</pre>
<p>clone this git repository and add it to your path:</p>
<pre>git clone git@github.com:SheffieldML/GPy.git ~/SheffieldML
echo 'PYTHONPATH=$PYTHONPATH:~/SheffieldML' &gt;&gt; ~/.bashrc
</pre>
</div>
<div id="compiling-documentation">
<h2>Compiling documentation:</h2>
<p>The documentation is stored in doc/ and is compiled with the Sphinx
Python documentation generator, and is written in the reStructuredText
format.</p>
<p>The Sphinx documentation is available here:
<a href="http://sphinx-doc.org/latest/contents.html" rel="nofollow">http://sphinx-doc.org/latest/contents.html</a></p>
<p><strong>Installing dependencies:</strong></p>
<p>To compile the documentation, first ensure that Sphinx is installed. On
Debian-based systems, this can be achieved as follows:</p>
<pre>sudo apt-get install python-pip
sudo pip install sphinx
</pre>
<p><strong>Compiling documentation:</strong></p>
<p>The documentation can be compiled as follows:</p>
<pre>cd doc
sphinx-apidoc -o source/ ../GPy/
make html
</pre>
<p>The HTML files are then stored in doc/build/html</p>
</div>
<div id="funding-acknowledgements">
<h2>Funding Acknowledgements</h2>
<p>Current support for the GPy software is coming through the following
projects.</p>

<p>Previous support for the GPy software came from the following projects:</p>
<ul>
<li><a href="http://staffwww.dcs.shef.ac.uk/people/N.Lawrence/projects/recombinant/" rel="nofollow">BBSRC Project No
BB/K011197/1</a>
“Linking recombinant gene sequence to protein product
manufacturability using CHO cell genomic resources”</li>
<li><a href="http://staffwww.dcs.shef.ac.uk/people/N.Lawrence/projects/biopredyn/" rel="nofollow">EU FP7-KBBE Project Ref
289434</a>
“From Data to Models: New Bioinformatics Methods and Tools for
Data-Driven Predictive Dynamic Modelling in Biotechnological
Applications”</li>
<li><a href="http://staffwww.dcs.shef.ac.uk/people/N.Lawrence/projects/iterative/" rel="nofollow">BBSRC Project No
BB/H018123/2</a>
“An iterative pipeline of computational modelling and experimental
design for uncovering gene regulatory networks in vertebrates”</li>
<li><a href="http://staffwww.dcs.shef.ac.uk/people/N.Lawrence/projects/synergy/" rel="nofollow">Erasysbio</a>
“SYNERGY: Systems approach to gene regulation biology through nuclear
receptors”</li>
</ul>
</div>


<a name="downloads"> </a>


<ul class="nodot">
  <li><strong>Downloads (All Versions):</strong></li>
  <li>
    <span>0</span> downloads in the last day
  </li>
  <li>
    <span>167</span> downloads in the last week
  </li>
  <li>
    <span>1266</span> downloads in the last month
  </li>
</ul>



<ul class="nodot">
 <li>
  <strong>Author:</strong>
  <span>`GPy Authors &lt;https://github.com/SheffieldML/GPy/graphs/contributors&gt;`__</span>
 </li>

 

 <li>
  <strong>Documentation:</strong>
  <a href="http://pythonhosted.org/GPy/">GPy package documentation</a>
 </li>


 <li>
  <strong>Home Page:</strong>
  
  <a href="http://sheffieldml.github.com/GPy/">http://sheffieldml.github.com/GPy/</a>
 </li>


 

 


 <li>
  <strong>Keywords:</strong>
  <span>machine-learning gaussian-processes kernels</span>
 </li>

 <li>
  <strong>License:</strong>
  
  
  <span>BSD 3-clause</span>
  
 </li>

 


 
 
 

 <li>
  <strong>Categories</strong>
  
 </li>



 

 

 

 

 


 <li>
  <strong>Package Index Owner:</strong>
  <span>walkingrandomly, maxz, sheffieldml</span>
 </li>

 <li>
  <strong>Package Index Maintainer:</strong>
  <span>walkingrandomly, maxz</span>
 </li>

 <li>
  <strong><a href="http://usefulinc.com/doap">DOAP</a> record:</strong>
  <a href="/pypi?:action=doap&amp;name=GPy&amp;version=0.9.1">GPy-0.9.1.xml</a>
 </li>

</ul>





            </div>


          </div></body></html>