<html><body><div><div id="the-northern-arizona-snp-pipeline-nasp">
<h2>The Northern Arizona SNP Pipeline (NASP)</h2>
<div id="overview">
<h3>OVERVIEW:</h3>
<p>NASP is a pipeline for analysis of genomic data. It is a suite of tools
meant to collect and report on statistically-relevant high-confidence
positions in a collection of genomes, with emphasis on variant
positions, especially single nucleotide polymorphisms (SNPs). NASP
expects some combination of files in FASTA, FASTQ, SAM, BAM, and VCF
format as input, and will produce output files also in
those formats. As NASP is a pipeline, it expects to link a set of
external tools (usually installed separately) to complete specific
analysis tasks.</p>
</div>
<div id="usage">
<h3>USAGE:</h3>
<p>Usage depends upon the installation method used on your system, and the
user interface you select. For standard installations with the
command-line interface, you would collect (or symbolically link) all of
your input files into a folder, and then invoke the command-line
interface from that folder. Expected usage for the command-line
interface is:</p>
<p><cite>nasp.py [output_folder]</cite></p>
<p>You will then be prompted to answer a few questions about your analysis.</p>
<p>Optionally, if you are re-running a previous analysis with the same (or
similar) options, you can pass in an xml-based configuration file (this
is written out to your output_folder after running the command-line
interface) using the format:</p>
<p><cite>nasp.py –config</cite></p>
</div>

<div id="dependencies">
<h3>DEPENDENCIES:</h3>
<p>For information about external tools that are required, or can be
utilized, and those versions that have been tested to work with NASP,
refer to the main NASP page (<a href="http://tgennorth.github.io/NASP/" rel="nofollow">http://tgennorth.github.io/NASP/</a>)</p>
</div>
<div id="license">
<h3>LICENSE:</h3>
<p>Copyright © The Translational Genomics Research Institute See the
included “LICENSE” document.</p>
</div>

</div>
</div></body></html>